stable_baselines3/dqn/prioritized_replay_buffer.py

```diff
@@ -0,0 +1,221 @@
+        :param done: Whether the episode was finished after the transition to be stored.
+        :param infos: Eventual information given by the environment.
+        """
+        # store transition index with maximum priority in sum tree
+        self.tree.add(self.max_priority, self.count)
+
+        # update counters
+        self.count = (self.count + 1) % self.buffer_size
+        self.real_size = min(self.buffer_size, self.real_size + 1)
+
+        # store transition in the buffer
+        super().add(obs, next_obs, action, reward, done, infos)
+
+    def sample(self, batch_size: int, env: Optional[VecNormalize] = None) -> ReplayBufferSamples:
+        """
+        Sample elements from the prioritized replay buffer.
+
+        :param batch_size: Number of element to sample
+        :param env:associated gym VecEnv
+            to normalize the observations/rewards when sampling
+        :return: a batch of sampled experiences from the buffer.
+        """
+        assert self.buffer_size >= batch_size, "The buffer contains less samples than the batch size requires."
+
+        sample_idxs, tree_idxs = [], []
+        priorities = th.empty(batch_size, 1, dtype=th.float)
+
+        # To sample a minibatch of size k, the range [0, p_total] is divided equally into k ranges.
+        # Next, a value is uniformly sampled from each range. Finally the transitions that correspond
+        # to each of these sampled values are retrieved from the tree.
+        segment = self.tree.p_total / batch_size
+        for i in range(batch_size):
+            # extremes of the current segment
+            a, b = segment * i, segment * (i + 1)
+
+            # uniformely sample a value from the current segment
+            cumsum = random.uniform(a, b)
+
+            # tree_idx is a index of a sample in the tree, needed further to update priorities
+            # sample_idx is a sample index in buffer, needed further to sample actual transitions
+            tree_idx, priority, sample_idx = self.tree.get(cumsum)
+
+            priorities[i] = priority
+            tree_idxs.append(tree_idx)
+            sample_idxs.append(int(sample_idx.item()))
+
+        # probability of sampling transition i as P(i) = p_i^alpha / \sum_{k} p_k^alpha
+        # where p_i > 0 is the priority of transition i.
+        probs = priorities / self.tree.p_total
+
+        # Importance sampling weights.
+        # All weights w_i were scaled so that max_i w_i = 1.
+        weights = (self.real_size * probs) ** -self.beta
+        weights = weights / weights.max()
+
+        env_indices = np.random.randint(0, high=self.n_envs, size=(len(sample_idxs),))
+
+        if self.optimize_memory_usage:
+            next_obs = self._normalize_obs(
+                self.observations[(np.array(sample_idxs) + 1) % self.buffer_size, env_indices, :], env
+            )
+        else:
+            next_obs = self._normalize_obs(self.next_observations[sample_idxs, env_indices, :], env)
```

Comment:
pytype thinks this value might None, which is apparently true (see `buffers.py` for solutions)

Reply:
`self.next_observations` might be None if `self.optimize_memory_usage` is True, but then we wouldn't enter the `else` at line 212.
The same thing happens in `buffers.py`, so I don't quite understand where the difference is

Reply:
Hello @araffin,
could you have a look at this (or the PR in general) when you have time?

After that I'll work on the tests, documentation etc. in order to complete the PR.
Cheers!

Reply:
I'm pretty busy right now (writing paper for end of the month) and I was waiting for the tests to pass, but I'll try to have a look beginning of October ;)